src/misc/coloring.c

```diff
@@ -129,6 +134,150 @@ static igraph_error_t igraph_i_vertex_coloring_greedy_cn(const igraph_t *graph,
     return IGRAPH_SUCCESS;
 }
 
+/* DSATUR: Choose vertices based on the number of adjacent colours, i.e. "saturation degree" */
+
+typedef struct {
+    igraph_integer_t saturation_degree;
+    igraph_integer_t edge_degree;
+} dsatur_t;
+
+static int dsatur_t_compare(const void* left, const void* right) {
+    const dsatur_t *left_d  = (dsatur_t*)left;
+    const dsatur_t *right_d = (dsatur_t*)right;
+    if (left_d->saturation_degree == right_d->saturation_degree) {
+        return left_d->edge_degree - right_d->edge_degree;
+    }
+    return left_d->saturation_degree - right_d->saturation_degree;
+}
+
+static igraph_bool_t is_color_used_by_neighbour(
+    const igraph_vector_int_t *colors, igraph_integer_t color,
+    const igraph_vector_int_t *neighbours
+) {
+    igraph_integer_t nbr_cnt = igraph_vector_int_size(neighbours);
+
+    for (igraph_integer_t nbr_indx = 0; nbr_indx < nbr_cnt; nbr_indx++) {
+        igraph_integer_t nbr = VECTOR(*neighbours)[nbr_indx];
+        if (VECTOR(*colors)[nbr] == color) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+static void igraph_i_dsatur_update_heap(
+    const igraph_adjlist_t *adjlist, igraph_gen2wheap_t *node_degrees_heap,
+    igraph_vector_int_t *neighbours, igraph_vector_int_t *colors,
+    igraph_integer_t color
+) {
+    igraph_gen2wheap_delete_max(node_degrees_heap);
+    dsatur_t color_metadata;
+    igraph_integer_t nbr_cnt = igraph_vector_int_size(neighbours);
+    for (igraph_integer_t nbr_indx = 0; nbr_indx < nbr_cnt; nbr_indx++) {
+        igraph_integer_t nbr = VECTOR(*neighbours)[nbr_indx];
+        if (!igraph_gen2wheap_has_elem(node_degrees_heap, nbr)) {
+            continue;
+        }
+        color_metadata = *((dsatur_t*)igraph_gen2wheap_get(node_degrees_heap, nbr));
+        if (!is_color_used_by_neighbour(colors, color, igraph_adjlist_get(adjlist, nbr))) {
+            color_metadata.saturation_degree++;
+        }
+        color_metadata.edge_degree--;
+        igraph_gen2wheap_modify(node_degrees_heap, nbr, &color_metadata);
+    }
+}
+
+static igraph_integer_t igraph_i_set_first_color(igraph_set_t *viable_colors) {
+    if (!igraph_set_contains(viable_colors, 0)) {
+        return 0;
+    }
+    igraph_integer_t state = 0, prev_element, next_element;
+    igraph_set_iterate(viable_colors, &state, &prev_element);
+    while (igraph_set_iterate(viable_colors, &state, &next_element)) {
+        if (next_element != prev_element + 1) {
+            return prev_element + 1;
+        }
+        prev_element = next_element;
+    }
+    return prev_element + 1;
```

Comment:
Here you are assuming that set elements are iterated over in increasing order. This is not guaranteed and cannot be relied on.

Why don't you use the sorted vector approach that I now suggested several times? Simply retrieve and sort the neighbour colours, then iterate through the list and find the first gap, just as here. It will be more efficient than building a set and it does not rely on internal implementation details. See the other heuristic for details.

Reply:
Sorry I'll do that now